velox/functions/prestosql/types/JsonType.cpp

```diff
@@ -509,292 +511,413 @@ void castToJsonFromRow(
-    }
+template <typename Input>
+struct CastFromJsonTypedImpl {
+  template <TypeKind kKind>
+  static simdjson::error_code apply(Input input, exec::GenericWriter& writer) {
+    return KindDispatcher<kKind>::apply(input, writer);
   }
-}
 
-template <>
-FOLLY_ALWAYS_INLINE void castFromJsonTyped<TypeKind::MAP>(
-    const folly::dynamic& object,
-    exec::GenericWriter& writer) {
-  auto& writerTyped = writer.castTo<Map<Any, Any>>();
-
-  for (const auto& pair : object.items()) {
-    // If casting to map of JSON values, nulls in map values should become the
-    // JSON text "null".
-    if (!isJsonType(writer.type()->childAt(1)) && pair.second.isNull()) {
-      auto& keyWriter = writerTyped.add_null();
-      VELOX_DYNAMIC_TYPE_DISPATCH(
-          castFromJsonTyped,
-          writer.type()->childAt(0)->kind(),
-          pair.first,
-          keyWriter);
-    } else {
-      auto writerPair = writerTyped.add_item();
-      VELOX_DYNAMIC_TYPE_DISPATCH(
-          castFromJsonTyped,
-          writer.type()->childAt(0)->kind(),
-          pair.first,
-          std::get<0>(writerPair));
-      VELOX_DYNAMIC_TYPE_DISPATCH(
-          castFromJsonTyped,
-          writer.type()->childAt(1)->kind(),
-          pair.second,
-          std::get<1>(writerPair));
+ private:
+  // Dummy is needed because full/explicit specialization is not allowed inside
+  // class.
+  template <TypeKind kKind, typename Dummy = void>
+  struct KindDispatcher {
+    static simdjson::error_code apply(Input, exec::GenericWriter&) {
+      VELOX_NYI(
+          "Casting from JSON to {} is not supported.", TypeTraits<kKind>::name);
     }
-  }
-}
-
-template <>
-FOLLY_ALWAYS_INLINE void castFromJsonTyped<TypeKind::ROW>(
-    const folly::dynamic& object,
-    exec::GenericWriter& writer) {
-  VELOX_USER_CHECK(
-      object.isArray() || object.isObject(),
-      "Only casting from JSON array or object to ROW is supported.");
-
-  auto& writerTyped = writer.castTo<DynamicRow>();
-
-  if (object.isArray()) {
-    VELOX_USER_CHECK_EQ(
-        writer.type()->size(),
-        object.size(),
-        "Cannot cast a JSON array of size {} to ROW with {} fields.",
-        object.size(),
-        writer.type()->size());
-
-    column_index_t i = 0;
-    for (auto it = object.begin(); it != object.end(); ++it, ++i) {
-      if (it->isNull()) {
-        writerTyped.set_null_at(i);
+  };
+
+  template <typename Dummy>
+  struct KindDispatcher<TypeKind::VARCHAR, Dummy> {
+    static simdjson::error_code apply(
+        Input value,
+        exec::GenericWriter& writer) {
+      SIMDJSON_ASSIGN_OR_RAISE(auto t, value.type());
+      std::string_view s;
+      if (isJsonType(writer.type())) {
+        SIMDJSON_ASSIGN_OR_RAISE(s, rawJson(value, t));
       } else {
-        VELOX_DYNAMIC_TYPE_DISPATCH(
-            castFromJsonTyped,
-            writer.type()->childAt(i)->kind(),
-            *it,
-            writerTyped.get_writer_at(i));
+        switch (t) {
```

Comment:
single-letter variables make code hard to read; what is t here? Consider renaming.